ipset: Rework the reconciler to use batch ops

pippolo84 opened the pull request:
Rework the ipset table to store a single address per entry. Consequently, rewrite the ipset reconciler to:

- use batch operations and `ipset restore`
- delete addresses only in DeleteBatch and Prune operations

Finally, introduce the concept of an ipset manager initializer, to delay the addresses pruning during agent restart, before the initial nodes listing from k8s or kvstore is completed.

Refer to the individual commit messages and the related issue for further details.

Fixes: https://github.com/cilium/cilium/issues/31537

joamaki commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -98,33 +81,41 @@ func (m *manager) RemoveFromIPSet(name string, addrs ...netip.Addr) {
 	ipset *ipset,
 	_ reconciler.Reconciler[*tables.IPSet], // needed to enforce the correct hive ordering
-) Manager {
-	db.RegisterTable(table)
```
```
func (m *manager) NewHandle() Handle {
  if m.started.Load() {
    panic(...)
  }
  m.startWG.Add(1)
  ...
}
// onStart is a oneshot async func
func (m *manager) onStart() {
  m.startWG.Wait() // wait for all handles to do initial population
  m.reconcilerLifecycle.Start(ctx) // start the reconciler
}

func (h *handle) InitDone() {
  h.manager.startWG.Done()
}
type Handle interface {
  AddToIPSet(...)
  InitDone()
}
```

pippolo84 committed:
iptables/ipset: Store a single addr per table entry

Rework the ipset table to store a single address per entry. This allows
to rewrite the reconciler in a more canonical way, instead of mixing
address insertions and removals in the Update operation.

In the current setup, the Update and Delete respectively add and remove
a single IP address in the related set.  The Prune operation performs a
full reconciliation for both IPv4 and IPv6 sets.

Since each table entry now stores a single address instead of an entire
set, the usage of immutable sets is replaced by a simpler netip.Addr
field.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables/ipset: Add initializer to delay entries deletion

The current ipset reconciler logic might remove ipset entries too
aggressively at startup, when the Cilium managed IP sets still contain
entries related to a previous agent run.  This may happen if a full
reconciliation is performed before the initial nodes listing (either
from k8s or kvstore) is complete.  This in turn may result in temporary
traffic disruption until the IP sets have been fully reloaded following
the completion of the initial nodes listing.

To avoid depending from k8s or the kvstore in the ipset cell to
explicitly wait for the listing, we rely on the handler pattern: the
ipset manager exposes a way to grab a handler (namely, an initializer)
that blocks the Prune operation until the initialization is explicitly
marked as completed.  The ipset manager consumer is in charge of
signalling the completion of the initialization and the start of the
full reconciliation.

The ipset manager waits on the initializers and as soon as all have been
marked as done triggers a full reconciliation to immediately perform a
Prune operation.

Since the Prune operation is always executed as soon as the initializers
have been marked as done, the full reconciliation interval is increased
to 30 minutes.

In a subsequent commit an ipset initializer is used by the node manager
to signal the possibility to safely perform address pruning.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 commented on pkg/datapath/iptables/ipset/cell.go:
```diff
@@ -60,7 +63,7 @@ func newReconcilerConfig(
 ) reconciler.Config[*tables.IPSetEntry] {
 	return reconciler.Config[*tables.IPSetEntry]{
-		FullReconcilationInterval: 10 * time.Second,
+		FullReconcilationInterval: 10 * time.Minute,
```
Increase to 30 min

pippolo84 commented on pkg/datapath/tables/ipset.go:
```diff
@@ -5,91 +5,65 @@ package tables
 )
 
-const IPSetTableName = "ipset"
+const IPSetEntryTableName = "ipset-entry"
```
Change to "ipsets"

pippolo84 committed:
iptables/ipset: Add a unit test for reconciler prune operation

Add a unit test to check that the ipset reconciler skips the Prune
operation up until all initializers have been marked as done.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables/ipset: Use batch operations

Performing an ipset list operation at each reconciler update can become
expensive in terms of memory, because it requires the allocation of a
set of addresses each time it is called. In a very large cluster or in a
large multi-cluster setup, the cardinality of this list can be pretty
high. When the cluster is experiencing node churn, and the
ipset.RemoveFromIPSet is called many times, this can lead to a very high
GC pressure, because we rely on `ipset list` in the reconciler Delete
operation to be sure that the IP set already exists.

To avoid that, reconcile the ipset changes using "ipset restore" in
order to add multiple IPs in one go and avoid the overhead of executing
ipset many times.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb/reconciler: Do not do WriteTxn on empty results

If the incremental round found no work to do we don't need to
acquire the WriteTxn. This can happen when objects are updated
but are not marked pending (e.g. changes are being made that
do not need to be reconciled).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

pippolo84 commented on pkg/datapath/iptables/ipset/ops.go:
```diff
@@ -103,19 +135,16 @@ func reconcile(
-			return fmt.Errorf("unable to add addr %s from ipset %s: %w", name, addr, err)
-		}
+	addBatch := map[string][]netip.Addr{name: toAdd.UnsortedList()}
+	if err := ipset.delBatch(ctx, addBatch); err != nil {
```
Should be `ipset.addBatch`

pippolo84 commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -263,16 +262,43 @@ func (i *ipset) add(ctx context.Context, name string, addr netip.Addr) error {
+	return i.exec(ctx, "ipset", "", args...)
+}
+
+func (i *ipset) runStdin(ctx context.Context, stdin string, args ...string) ([]byte, error) {
```
This seems unused

giorio94 commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -156,6 +188,11 @@ func newIPSetManager(
+	// wait for all initializers to complete before finalizing manager
+	// initialization and allowing prune operations in the ipset reconciler
+	m.started.Store(true)
+	m.startWG.Wait()
```
I guess this could be moved below the ipset cleanup step, as that's independent of node synchronization.

giorio94 commented on pkg/node/manager/manager.go:
```diff
@@ -827,6 +829,13 @@ func (m *manager) NodeDeleted(n nodeTypes.Node) {
+// NodeSync signals the manager that the initial nodes listing (either from k8s or kvstore)
+// has been completed. This allows the manager to initiate the deletion of possible stale entries
+// in the kernel IP sets managed by Cilium.
+func (m *manager) NodeSync() {
+	m.ipsetInitializer.InitDone()
+}
```
`NodeSync` may be called twice if the kvstore is enabled (once when the informer syncs, the other when the kvstore syncs). However, `InitDone` is implemented on top of a wait group, which panics if it turns negative.

giorio94 replied:
Additionally, this logic is not accounting for the clustermesh case. You can retrieve whether clustermesh is synchronized using `(*ClusterMesh).NodesSynced()`.

pippolo84 replied:
Nice catch! Switched to an int counter with a `sync.Cond`, so we can check for counter to be equal to 0 or negative. Also, added an Invoke in clustermesh cell to wait for external nodes sync.

giorio94 replied:
Hmm, I only had a quick look, but I'm still not convinced this would work as expected.

In the the kvstore + clustermesh case, the following sequence may occur:

* Node manager + clustermesh take the initializer -> counter = 2
* CRD synchronization -> InitDone -> counter = 1
* kvstore synchronization -> InitDone -> counter = 0
* The prune operation is executed, even though clustermesh has not yet synchronized. 

I think we should either make sure that `InitDone` is called exactly once, or make it idempotent (i.e., a no-op after the first execution).

As a side note, you could use a StoppableWaitGroup instead of the custom counter + `sync.Cond` logic.

pippolo84 replied:
Yeah, the fact that we can flip back and forth from kvstore to crd is not playing nice with the initializers logic.

> I think we should either make sure that InitDone is called exactly once, or make it idempotent (i.e., a no-op after the first execution).

I think the first requires an ad-hoc solution for the kvstore/crd case, it is probably better to make the `InitDone` idempotent, so to keep having one "signal" for each "initializer" taken from the manager.

I'll try with the `StoppableWaitGroup`, thanks for the heads up!

giorio94 commented on pkg/datapath/iptables/ipset/ops.go:
```diff
@@ -30,37 +31,61 @@ type ops struct {
 
-func (ops *ops) Update(ctx context.Context, _ statedb.ReadTxn, entry *tables.IPSetEntry, changed *bool) error {
+// DeleteBatch implements reconciler.BatchOperations.
+func (ops *ops) DeleteBatch(ctx context.Context, txn statedb.ReadTxn, batch []reconciler.BatchEntry[*tables.IPSetEntry]) {
```
Nit. It seems a bit more natural if UpdateBatch comes before DeleteBatch (and the diff would also improve a bit).

giorio94 commented on pkg/datapath/iptables/ipset/ops.go:
```diff
@@ -30,37 +31,61 @@ type ops struct {
-	if err := ops.ipset.add(ctx, entry.Name, entry.Addr); err != nil {
-		return err
+// UpdateBatch implements reconciler.BatchOperations.
+func (ops *ops) UpdateBatch(ctx context.Context, txn statedb.ReadTxn, batch []reconciler.BatchEntry[*tables.IPSetEntry]) {
```
Differently from before, both `UpdateBatch` and `DeleteBatch` no longer check whether the ipset actually exists before attempting the operations. It looks reasonable to me, especially performance-wise, just making sure that the change was intentional.

pippolo84 replied:
Yep, that is intentional and it is meant to further improve the performance removing the exec to run `ipset create`. In case of removed sets, we should be able to recover with ta full reconciliation.

giorio94 commented on pkg/statedb/reconciler/reconciler.go:
```diff
@@ -107,12 +107,16 @@ func WaitForReconciliation[Obj any](ctx context.Context, db *statedb.DB, table s
 }
 
 func (r *reconciler[Obj]) loop(ctx context.Context, health cell.HealthReporter) error {
+
```
Nit. Should be likely reverted.

giorio94 commented on pkg/statedb/reconciler/reconciler.go:
```diff
@@ -107,12 +107,16 @@ func WaitForReconciliation[Obj any](ctx context.Context, db *statedb.DB, table s
 		defer fullReconTicker.Stop()
 		fullReconTickerChan = fullReconTicker.C
 	}
+	if r.Config.RateLimiter != nil {
+		defer r.Config.RateLimiter.Stop()
+	}
```
Nit. May be worth to move it to a separate commit, as it appears an unrelated fix.

giorio94 reviewed (changes requested):
Thanks!

A couple of comments inline about waiting for clustermesh synchronization as well, and a possible panic due to double sync. And a few minor nits.

Concerning the node synchronization logic, it may be worth considering that we'll likely need something very similar for WireGuard to address https://github.com/cilium/cilium/issues/31985, to ensure that the current logic is extensible enough to address that case as well.

pippolo84 committed:
reconciler: Stop the rate limiter goroutine if needed

Stop the rate limiter goroutine during reconciler termination if rate
limiting was enabled in the configuration.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
clustermesh: Delay ipset reconciliation after nodes synchronization

When clustermesh is enabled, the nodes manager receives events about all
the remote cluster nodes. Those events are handled just as nodes events
coming from the local cluster, so the remote nodes IPs are added to the
kernel IP sets.

To avoid pruning remote nodes addresses in kernel IP sets too early
after an agent restart, clustermesh module takes an initializer handler
to the ipset manager, delayiing the stale addresses deletion until the
listing of all the remote clusters nodes is completed.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

giorio94 commented on daemon/cmd/status_test.go:
```diff
@@ -36,14 +38,36 @@ var (
 func (g *GetNodesSuite) SetUpSuite(c *C) {
 	var err error
-	nm, err = manager.New(fakeConfig, nil, nil, nil, manager.NewNodeMetrics(), cell.TestScope())
+	nm, err = manager.New(fakeConfig, nil, newFakeIPSetMock(), nil, manager.NewNodeMetrics(), cell.TestScope())
```
Any reason for not using the fake type in `pkg/datapath/fake/types/ipset.go`?

pippolo84 replied:
I thought we will end up in an import loop but it seems I was wrong. Fixed.

giorio94 commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -50,17 +51,13 @@ type Initializer interface {
 type initializer struct {
-	cond    *sync.Cond
-	counter *int
+	once sync.Once
+	wg   *lock.StoppableWaitGroup
```
Should be moved to the `iptables/ipset: Add initializer to delay entries deletion` commit. The same applies for the other changes in this file.

pippolo84 replied:
Oops, I've messed up the rebase. I'm gonna fix it.

giorio94 replied:
Thanks!

giorio94 commented on pkg/clustermesh/ipsetsyncer.go:
```diff
@@ -0,0 +1,45 @@
+	jg.Add(job.OneShot("clustermesh-ipset-syncer", func(ctx context.Context, _ cell.HealthReporter) error {
+		// wait for initial nodes listing from all remote clusters
+		// before allowing stale ipset entries deletion
+		cm.NodesSynced(ctx)
```
Nit. It is likely better to check the possible error, and return in that case. The only possible error is if the context gets closed, but in that case we don't want to accidentally trigger the pruning step due to race conditions (although theoretically the corresponding context should be canceled as well).

giorio94 reviewed (commented):
Thanks! A couple of minor comments, then I'll do the final pass.

pippolo84 committed:
node/manager: Add method to signal completion of initial nodes listing

The Cilium managed IP sets contain the node addresses for which we do
not SNAT the outgoing traffic.  During an agent restart, while we are
waiting for the initial nodes listing to complete, we don't want to
remove addresses too early, because those addresses may still be related
to existing nodes not yet seen in the in-progress listing.

In order to avoid temporary pod traffic disruption, the node manager
takes an ipset initializer reference and mark it as done only after the
initial listing, from either k8s or kvstore, is complete.
This in turn triggers a full IP sets reconciliation, to safely remove
stale addresses that are still in the sets.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

giorio94 reviewed (approved):
Thanks!
